src/librustc_mir/transform/generator.rs

```diff
@@ -1166,7 +1188,28 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
 
         // We rename RETURN_PLACE which has type mir.return_ty to new_ret_local
         // RETURN_PLACE then is a fresh unused local with type ret_ty.
-        let new_ret_local = replace_result_variable(ret_ty, body, tcx);
+        let new_ret_local = replace_local(RETURN_PLACE, ret_ty, body, tcx);
+
+        // We also replace the resume argument and insert an `Assign`.
+        // This is needed because the resume argument might be live across a `yield`, and the
+        // transform assumes that any local live across a `yield` is assigned to before that.
```

Comment:
Where does the transform assume this?

Reply:
It doesn't explicitly assume it, but the way it rewrites any mention of locals stored in the generator state doesn't work with an argument that is passed in (and initialized) from the outside. There's never a store to `_2` because it's already initialized, so there won't be any store to its slot in the generator state either, and then the first mention after the yield will read from the uninitialized slot.

Reply:
I think you should rephrase the comment so it says we need to introduce a store to the resume argument so that such a store to the generator state will exist if it's live across a yield.

Reply:
Yeah, good idea. I've pushed an updated comment.